File: Go/remove_linked_list_elements.go
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

func removeElements(head *ListNode, val int) *ListNode {
	// когда первый элемент списка равен значению val и его нужно удалить.
	dummy := &ListNode{Next: head}
	cur := dummy

	for cur.Next != nil && cur.Next.Val == val {
		if val == cur.Next.Val {
			cur.Next = cur.Next.Next
		} else {
			cur = cur.Next
		}
	}
	return dummy.Next
}

File: C/remove_linked_list_elements.c
#include <stddef.h>  // NULL
#include <stdlib.h>  // malloc


struct ListNode {
    int val; 
    struct ListNode *next;
};

struct ListNode* removeElements(struct ListNode* head, int val) {
    struct ListNode* dummy = malloc(sizeof(struct ListNode));
    dummy->next = head;
    struct ListNode* cur = dummy;

    while (cur != NULL && cur->next != NULL) {
        if (val == cur->next->val) {
            cur->next = cur->next->next;
        } else {
            cur = cur->next;
        }
    }
    return dummy->next;
};